roody/doxfiles/man/man3/RoodyXML.3
.TH "RoodyXML" 3 "Thu Apr 26 2012" "Roody" \" -*- nroff -*-
.ad l
.nh
.SH NAME
RoodyXML \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <RoodyXML.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "int \fBFindNode\fP (\fBPMXML_NODE\fP parent, int startIndex, const char *name)"
.br
.ti -1c
.RI "\fBPMXML_NODE\fP \fBGetNode\fP (\fBPMXML_NODE\fP parent, int index)"
.br
.ti -1c
.RI "std::string \fBGetNodeString\fP (\fBPMXML_NODE\fP node, int index)"
.br
.ti -1c
.RI "\fBRoodyXML\fP ()"
.br
.ti -1c
.RI "virtual \fB~RoodyXML\fP ()"
.br
.ti -1c
.RI "std::ofstream & \fBOpenFileForWrite\fP (char const *)"
.br
.ti -1c
.RI "std::string \fBEncode\fP (std::string const &)"
.br
.ti -1c
.RI "bool \fBOpenFileForRead\fP (char const *)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "unsigned int \fBlevel_\fP"
.br
.ti -1c
.RI "bool \fBelementIsOpen_\fP"
.br
.ti -1c
.RI "std::vector< std::string > \fBnames_\fP"
.br
.ti -1c
.RI "std::ofstream \fBoutputFile_\fP"
.br
.ti -1c
.RI "char const * \fBencoding_\fP"
.br
.ti -1c
.RI "char const * \fBversion_\fP"
.br
.ti -1c
.RI "\fBPMXML_NODE\fP \fBdocumentNode_\fP"
.br
.ti -1c
.RI "\fBPMXML_NODE\fP \fBroodyNode_\fP"
.br
.ti -1c
.RI "\fBPMXML_NODE\fP \fBhistParent_\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 21 of file RoodyXML.h.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "RoodyXML::RoodyXML ()"
.PP
Definition at line 14 of file RoodyXML.cxx.
.SS "RoodyXML::~RoodyXML ()\fC [virtual]\fP"
.PP
Definition at line 20 of file RoodyXML.cxx.
.SH "Member Function Documentation"
.PP 
.SS "std::string RoodyXML::Encode (std::string const &src)"
.PP
Definition at line 59 of file RoodyXML.cxx.
.PP
Referenced by Roody::SaveFile(), and Roody::SaveFilePadContents().
.SS "int RoodyXML::FindNode (\fBPMXML_NODE\fPparent, intstartIndex, const char *name)"
.PP
Definition at line 108 of file RoodyXML.cxx.
.PP
Referenced by OpenFileForRead(), Roody::RestoreFile(), and Roody::RestoreObjects().
.SS "\fBPMXML_NODE\fP RoodyXML::GetNode (\fBPMXML_NODE\fPparent, intindex)"
.PP
Definition at line 118 of file RoodyXML.cxx.
.PP
Referenced by Roody::RestoreFile(), and Roody::RestoreObjects().
.SS "std::string RoodyXML::GetNodeString (\fBPMXML_NODE\fPnode, intindex)"
.PP
Definition at line 126 of file RoodyXML.cxx.
.PP
Referenced by Roody::RestoreFile(), and Roody::RestoreObjects().
.SS "bool RoodyXML::OpenFileForRead (char const *file)"
.PP
Definition at line 89 of file RoodyXML.cxx.
.PP
Referenced by Roody::RestoreFile().
.SS "std::ofstream & RoodyXML::OpenFileForWrite (char const *file)"
.PP
Definition at line 23 of file RoodyXML.cxx.
.PP
Referenced by Roody::SaveFile().
.SH "Member Data Documentation"
.PP 
.SS "\fBPMXML_NODE\fP \fBRoodyXML::documentNode_\fP"
.PP
Definition at line 30 of file RoodyXML.h.
.PP
Referenced by OpenFileForRead().
.SS "bool \fBRoodyXML::elementIsOpen_\fP"
.PP
Definition at line 25 of file RoodyXML.h.
.SS "char const* \fBRoodyXML::encoding_\fP"
.PP
Definition at line 28 of file RoodyXML.h.
.PP
Referenced by OpenFileForWrite(), and RoodyXML().
.SS "\fBPMXML_NODE\fP \fBRoodyXML::histParent_\fP"
.PP
Definition at line 30 of file RoodyXML.h.
.SS "unsigned int \fBRoodyXML::level_\fP"
.PP
Definition at line 24 of file RoodyXML.h.
.SS "std::vector<std::string> \fBRoodyXML::names_\fP"
.PP
Definition at line 26 of file RoodyXML.h.
.SS "std::ofstream \fBRoodyXML::outputFile_\fP"
.PP
Definition at line 27 of file RoodyXML.h.
.PP
Referenced by OpenFileForWrite().
.SS "\fBPMXML_NODE\fP \fBRoodyXML::roodyNode_\fP"
.PP
Definition at line 30 of file RoodyXML.h.
.PP
Referenced by OpenFileForRead(), and Roody::RestoreFile().
.SS "char const* \fBRoodyXML::version_\fP"
.PP
Definition at line 29 of file RoodyXML.h.
.PP
Referenced by OpenFileForWrite(), and RoodyXML().

.SH "Author"
.PP 
Generated automatically by Doxygen for Roody from the source code.


roody/src/RoodyXML.cxx
//
// RoodyXML.cxx
//
// $Id: RoodyXML.cxx 234 2010-04-21 00:15:43Z olchansk $
//

#include <iostream>
#include <ios>
#include <time.h>
#include <string.h>

#include "RoodyXML.h"

RoodyXML::RoodyXML()
{
  version_ = "1.0";
  encoding_ = "ISO-8859-1";
}

RoodyXML::~RoodyXML()
{}

std::ofstream &RoodyXML::OpenFileForWrite( char const *file )
{
  if( outputFile_.is_open() )outputFile_.close();
  std::ios_base::openmode mode = std::ios_base::out|std::ios_base::trunc;
  outputFile_.open( file, mode );
  //
  if( outputFile_ )
  {
    outputFile_ << "<?xml version=\"" << version_
                << "\" encoding=\"" << encoding_ << "\"?>\n";
    /*
      << "<!DOCTYPE roody [\n"
      << "  <!ELEMENT roody (online?,file*,graphics?,canvas*,group*)>\n"
      << "  <!ELEMENT online (host,port?)>\n"
      << "  <!ELEMENT graphics (zones?,refresh?)>\n"
      << "  <!ELEMENT zones (#PCDATA)>\n"
      << "  <!ELEMENT refresh (#PCDATA)>\n"
      << "  <!ELEMENT canvas (zones?,histogram*)>\n"
      << "  <!ELEMENT group (name,histogram*)>\n"
      << "  <!ELEMENT name (#PCDATA)>\n"
      << "  <!ELEMENT histogram (name,source)>\n"
      << "  <!ELEMENT source (#PCDATA)>\n"
      << "  <!ELEMENT host (#PCDATA)>\n"
      << "  <!ELEMENT port (#PCDATA)>\n"
      << "  <!ELEMENT file (#PCDATA)>\n"
      << "]>\n";
    */
    time_t now;
    time( &now );
    std::string temp( ctime(&now) );
    temp.erase( temp.begin()+temp.size()-1 );
    outputFile_ << "<!-- created by Roody on " << temp << " -->\n";
  }
  return outputFile_;
}

std::string RoodyXML::Encode( const std::string &src )
{
  std::string result;
  std::size_t length = src.size();
  for( std::size_t i=0; i<length; ++i )
  {
    switch ( src[i] )
    {
      case '<':
        result += "&lt;";
        break;
      case '>':
        result += "&gt;";
        break;
      case '&':
        result += "&amp;";
        break;
      case '\"':
        result += "&quot;";
        break;
      case '\'':
        result += "&apos;";
        break;
      default:
        result += src[i];
    }
  }  
  return result;
}

bool RoodyXML::OpenFileForRead( const char *file )
{
  char error[256];
  documentNode_ = mxml_parse_file( const_cast<char*>(file), error, static_cast<int>(sizeof(error)) );
  if( !documentNode_ )
  {
    std::cerr << error << std::endl;
    return false;
  }
  int i = FindNode( documentNode_, 0, "roody" );
  if( i < 0 )
  {
    std::cerr << "roody node not found in file: \"" << file << "\"" << std::endl;
    return false;
  }
  roodyNode_ = &documentNode_->child[i];
  return true;
}

int RoodyXML::FindNode( PMXML_NODE parent, int startIndex, char const *nameToFind )
{
  int numberOfSiblings = parent->n_children;
  for( int i=startIndex; i<numberOfSiblings; ++i )
  {
    if( !strcmp(parent->child[i].name,nameToFind) )return i;
  }
  return -1;
}

PMXML_NODE RoodyXML::GetNode(PMXML_NODE parent, int index)
{
  if (parent==0 || index<0)
    return 0;

  return &parent->child[index];
}

std::string RoodyXML::GetNodeString(PMXML_NODE parent, int index)
{
  if (parent==0 || index<0)
    return "";

  const char* s = parent->child[index].value;
  if (!s)
    s = "";

  return s;
}

/*
void RoodyXML::PrintTree( int level )
{
  PrintTree( level, documentNode_ );
}

void RoodyXML::PrintTree( int level, PMXML_NODE current )
{
  while( current )
  {
    for( int i=0; i<level; ++i )std::cout << "  ";
    std::cout << "current " << current->name << ", type " << current->node_type
              << ", children " << (void*)current << "\n";
    if( current->children )PrintTree( level+1, current->children );
    current = current->next;
  }
}
*/
// end of file
